analyzer/queries/queries.go

```diff
@@ -850,4 +876,108 @@ var (
       source_files = $5
     WHERE
       runtime = $1 AND contract_address = $2`
+
+	RuntimeEvmVerifiedContractTxs = `
+    WITH abi_contracts AS (
+      SELECT 
+        runtime,
+        contract_address AS addr,
+        abi
+      FROM chain.evm_contracts
+      WHERE
+        runtime = $1 AND abi IS NOT NULL
+    )
+    SELECT 
+      abi_contracts.addr,
+      abi_contracts.abi,
+      txs.tx_hash,
+      txs.body->'data',
+      txs.error_message_raw
+    FROM abi_contracts 
+    JOIN chain.runtime_transactions as txs ON
+      txs.runtime = abi_contracts.runtime AND
+      txs.to = abi_contracts.addr AND
+      txs.method = 'evm.Call' -- note: does not include evm.Create txs; their payload is never encrypted.
+    WHERE
+      body IS NOT NULL AND
+      abi_parsed_at IS NULL
+    ORDER BY addr
+    LIMIT $2`
+
+	RuntimeEvmVerifiedContractTxsCount = `
+    WITH abi_contracts AS (
+      SELECT 
+        runtime,
+        contract_address AS addr
+      FROM chain.evm_contracts
+      WHERE
+        runtime = $1 AND abi IS NOT NULL
+    )
+    SELECT COUNT(*)
+    FROM abi_contracts 
+    JOIN chain.runtime_transactions as txs ON
+      txs.runtime = abi_contracts.runtime AND
+      txs.to = abi_contracts.addr AND
+      txs.method = 'evm.Call' -- note: does not include evm.Create txs; their payload is never encrypted.
+    WHERE
+      body IS NOT NULL AND
+      abi_parsed_at IS NULL`
+
+	RuntimeEvmVerifiedContractEvents = `
+    WITH abi_contracts AS (
+      SELECT
+        runtime,
+        contract_address AS addr,
+        abi
+      FROM chain.evm_contracts
+      WHERE
+        runtime = $1 AND
+        abi IS NOT NULL
+    )
+    SELECT
+      abi_contracts.addr,
+      abi_contracts.abi,
+      txs.round,
+      txs.tx_index,
+      evs.body
+    FROM abi_contracts
+    JOIN chain.runtime_transactions as txs ON
+      txs.runtime = abi_contracts.runtime AND
+      txs.to = abi_contracts.addr AND
+      txs.method = 'evm.Call'
+    JOIN chain.runtime_events as evs ON
+      evs.runtime = abi_contracts.runtime AND
+      evs.tx_hash = txs.tx_hash
```

Comment:
Based on today's slack conversation with Warren, we actually want to avoid joining on the txs because the contract that was the target of tx is not necessarily also the contract that emitted the event. The contract that emitted the event is in the event's `body.address`. So something like

```suggestion
    FROM abi_contracts
    JOIN chain.runtime_events as evs ON
      evs.type = 'evm.log' AND
      evs.runtime = abi_contracts.runtime AND
      decode(evs.body->>'address', 'base64') = eth_preimage(abi_contracts.contract_address)
```

That might require an index on body.address (IDK if it's needed, and/or if it would help), or maybe even a [generated column](https://www.postgresql.org/docs/current/ddl-generated-columns.html) for `decode(evs.body->>'address', 'base64')` and then an index on top of _that_. See how it goes with the simplest approach first :)

Reply:
Hrm it's slow, even with a generated column. Seems like the `eth_preimage` fn is the bottleneck - will try to just join over address_preimages instead
```
oasisindexer=*> explain analyze
WITH abi_contracts AS (
      SELECT
        runtime,
        contract_address AS addr,
        abi
      FROM chain.evm_contracts
      WHERE
        runtime = 'emerald' AND
        abi IS NOT NULL
    )
    SELECT
      abi_contracts.addr,
      abi_contracts.abi,
      evs.round,
      evs.tx_index,
      evs.body
    FROM abi_contracts
    JOIN chain.runtime_events as evs ON
      evs.type = 'evm.log' AND
      evs.runtime = abi_contracts.runtime AND
      evm_contract_address_eth = eth_preimage(abi_contracts.addr)
    LIMIT 1000;
                                                                  QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.00..181689.03 rows=1000 width=982) (actual time=42717.383..61984.504 rows=197 loops=1)
   ->  Nested Loop  (cost=0.00..2232049.74 rows=12285 width=982) (actual time=42717.381..61984.447 rows=197 loops=1)
         Join Filter: (evs.evm_contract_address_eth = eth_preimage(evm_contracts.contract_address))
         Rows Removed by Join Filter: 326507
         ->  Seq Scan on runtime_events evs  (cost=0.00..1580270.46 rows=614246 width=215) (actual time=15.804..55033.975 rows=46672 loops=1)
               Filter: ((runtime = 'emerald'::runtime) AND (type = 'evm.log'::text))
               Rows Removed by Filter: 20543893
         ->  Materialize  (cost=0.00..678.53 rows=4 width=807) (actual time=0.000..0.001 rows=7 loops=46672)
               ->  Seq Scan on evm_contracts  (cost=0.00..678.51 rows=4 width=807) (actual time=1.088..2.911 rows=7 loops=1)
                     Filter: ((abi IS NOT NULL) AND (runtime = 'emerald'::runtime))
                     Rows Removed by Filter: 7988
 Planning Time: 3.994 ms
 Execution Time: 61984.703 ms
(13 rows)
```

Reply:
:+1: is simple to use but not very efficient. That said, I thought it wouldn't matter here because we have so few verified contracts. But the query plan doesn't choose to go contract by contract, trying to find matching events for each contract :(

Reply:
Updated to join on `chain.address_preimages`. Also added an index on `chain.runtime_events (runtime, type)`, which makes the query performant enough.
```
oasisindexer=*> explain analyze
WITH abi_contracts AS (
      SELECT
        runtime,
        contract_address AS addr,
        abi
      FROM chain.evm_contracts
      WHERE
        runtime = 'emerald' AND
        abi IS NOT NULL
    )
    SELECT
      abi_contracts.addr,
      abi_contracts.abi,
      evs.round,
      evs.tx_index,
      evs.body
    FROM abi_contracts
    join chain.address_preimages as preimages on abi_contracts.addr = preimages.address JOIN chain.runtime_events as evs ON
      evs.type = 'evm.log' AND
      evs.runtime = abi_contracts.runtime AND
      decode(body->>'address', 'base64') = preimages.address_data
    LIMIT 1000;
                                                                                     QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=10154.37..1170949.61 rows=114 width=982) (actual time=257.164..2266.097 rows=197 loops=1)
   ->  Gather  (cost=10154.37..1170949.61 rows=114 width=982) (actual time=257.162..2266.052 rows=197 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Hash Join  (cost=9154.37..1169938.21 rows=48 width=982) (actual time=1015.169..2243.789 rows=66 loops=3)
               Hash Cond: (decode((evs.body ->> 'address'::text), 'base64'::text) = preimages.address_data)
               ->  Parallel Bitmap Heap Scan on runtime_events evs  (cost=8442.09..1167623.43 rows=256322 width=183) (actual time=5.136..2204.427 rows=15564 loops=3)
                     Recheck Cond: ((runtime = 'emerald'::runtime) AND (type = 'evm.log'::text))
                     Heap Blocks: exact=7413
                     ->  Bitmap Index Scan on ix_runtime_events_type  (cost=0.00..8288.29 rows=615173 width=0) (actual time=9.769..9.769 rows=46692 loops=1)
                           Index Cond: ((runtime = 'emerald'::runtime) AND (type = 'evm.log'::text))
               ->  Hash  (cost=712.23..712.23 rows=4 width=828) (actual time=5.272..5.274 rows=7 loops=3)
                     Buckets: 1024  Batches: 1  Memory Usage: 14kB
                     ->  Nested Loop  (cost=0.41..712.23 rows=4 width=828) (actual time=1.554..5.236 rows=7 loops=3)
                           ->  Seq Scan on evm_contracts  (cost=0.00..678.51 rows=4 width=807) (actual time=1.501..5.043 rows=7 loops=3)
                                 Filter: ((abi IS NOT NULL) AND (runtime = 'emerald'::runtime))
                                 Rows Removed by Filter: 7988
                           ->  Index Scan using address_preimages_pkey on address_preimages preimages  (cost=0.41..8.43 rows=1 width=68) (actual time=0.024..0.024 rows=1 loops=21)
                                 Index Cond: ((address)::text = (evm_contracts.contract_address)::text)
 Planning Time: 1.216 ms
 Execution Time: 2266.211 ms
(21 rows)
```

Side note: The generated column for the event address ended up being not needed, but will keep in mind for future performance testing.